src/allmydata/web/unlinked.py

```diff
@@ -59,34 +73,70 @@ def _done(upload_results, redir_to):
     return d
 
 
-class UploadResultsPage(status.UploadResultsRendererMixin, rend.Page):
+class UploadResultsPage(MultiFormatResource):
     """'POST /uri', to create an unlinked file."""
-    docFactory = getxmlfile("upload-results.xhtml")
 
     def __init__(self, upload_results):
-        rend.Page.__init__(self)
-        self.results = upload_results
+        """
+        :param IUploadResults upload_results: stats provider.
+        """
+        super(UploadResultsPage, self).__init__()
+        self._upload_results = upload_results
+
+    def render_HTML(self, req):
+        elem = UploadResultsElement(self._upload_results)
+        return renderElement(req, elem)
+
+    # This is weird but necessary because:
+    #
+    #  1. MultiFormatResource.render() uses argument "t" to figure out
+    #     its output format.
+    #
+    #  2. Upload request is of the form "POST /uri?t=upload&file=newfile".
+    #     See URIHandler.render_POST().
+    #
+    # MultiFormatResource.render() looks up "t" argument, which in
+    # this case has the value "upload", and then it would look for a
+    # render_UPLOAD() method.
+    #
+    # We could change upload request to use more descriptive names
+    # that do not cause name collisions like this.  That should be a
+    # separate change though.
+    render_UPLOAD = render_HTML
+
```

Comment:
All this could go away if it was just a `Resource`, right?

Reply:
Indeed!  All that is gone now, in 88d3fa7.  Thank you for pointing that out!